Update `rdf-canonize`, node versions, and more.

davidlehn opened the pull request.

davidlehn committed:
Drop support for Node.js < 18.

davidlehn committed:
Update to `rdf-canonize@4`.

- **BREAKING**: See the `rdf-canonize` 4.0.0 changelog for **important**
  changes and upgrade notes.
- Update to handle different RDF/JS dataset `BlankNode` format.
- Enable pass through of numerous possible `rdf-canonize` options in a
  `canonize()` `canonizeOptions` parameter.
- Update to use `rdf-canonize` options.
- The `URDNA2015` default algorithm has been changed to `RDFC-1.0` from
  `rdf-canon`.
- Complexity control defaults `maxWorkFactor` or `maxDeepIterations` may
  need to be adjusted to process graphs with certain blank node
  constructs.
- A `signal` option is available to use an `AbortSignal` to limit
  resource usage.
- The internal digest algorithm can be changed.

davidlehn committed:
Remove support for `application/nquads` alias.

davidlehn committed:
Update dependencies.

gabboman commented:
This pr fixes some security issues. Would be good to update the http library too to the 4.11 version.
the http library on the 3.xx branch has a dependency on udincy or something like that wich leaks memory.

Is there any way of helping with this or testing the beta branch with the fixes?

gabboman commented:
update: I tried myself installing just the newer version of the httpclient on my project and that caused an infinite loop in some cases. This may need some investigation and just updating dependencies may not just cut it

RinkeHoekstra commented:
Would be great if this change could be merged!

gabboman commented:
ignore my previous comment the library is safe was one of my changes that had an whoopsie

davidlehn committed:
Update actions.

- Use Node.js 22.x for most jobs.
- Rename workflow extension to .yaml.

davidlehn committed:
Update development dependencies.

- Update development dependencies.
- Update karma testing.
  - Remove older fixes in favor of more default behavior.
- Update bundle build.
  - Use newer corejs version.
  - Build with modern browserslist defaults and no IE support.
  - Support for older browsers requires a custom build.

davidlehn committed:
Update dependencies.

davidlehn committed:
Support and test Node.js 24.x.

davidlehn committed:
Refactor test framework.

- Test runtime loads test files from a web server.
- Allows testing of manifests on remote web servers.
- Trading off some performance to align node and browser testing.
- Moves some test setup code into config data and manifest.

davidlehn committed:
Fix browserify env vars.

davidlehn committed:
Fix test fromRdf#t0027.

davidlehn committed:
Fix conditional.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

davidlehn commented on lib/fromRdf.js:
```diff
@@ -350,10 +350,12 @@ function _RDFToObject(o, useNativeTypes, rdfDirection, options) {
+        } else if(rval['@value'] === 'false' || rval['@value'] === '0') {
           rval['@value'] = false;
+        } else if(rval['@value'] === 'True' || rval['@value'] === 'False') {
+          rval['@type'] = type;
```
> Thinking about it more, you should not get '1' or '0' from RDF.

Supporting those was part of why the newer test exists.  What did you mean?

> This should simply be an else.

Yes, I got confused.  Made a test suite PR to help clarify and handle another similar case that this code would fail on.
https://github.com/w3c/json-ld-api/pull/669

filip26 replied:
Yes, it's confusing but please keep the test as is, see [https://github.com/w3c/json-ld-api/pull/669#issuecomment-3461754481](https://github.com/w3c/json-ld-api/pull/669#issuecomment-3461754481).

Your fix is almost OK, just the plain `else` and it should work.

davidlehn committed:
Fix useNativeTypes handling.

- Fix previous useNativeTypes change.
- Refactor to make code easier to follow.
  - Performance change not tested.
- Handle another case for integeres similar to that from fromRdf#t0027.
  - Additions proposed for API test suite.

davidlehn committed:
Update dev dependencies.

davidlehn committed:
Use `rdf-canonize@5`.

- Update notes about removed `rdf-canonize-native` support.

davidlehn committed:
Update workflow.

- Update workflow name.
- Run checks on pull requests.
- Fix zizmor reported issues.
- Update action versions.
